newrelic-agent/src/main/java/com/newrelic/agent/AgentImpl.java

```diff
@@ -157,19 +164,58 @@ public TraceMetadata getTraceMetadata() {
 
         TraceMetadata traceMetadata = getTraceMetadata();
         String traceId = traceMetadata.getTraceId();
-        linkingMetadata.put("trace.id", traceId);
+        linkingMetadata.put(TRACE_ID, traceId);
 
         String spanId = traceMetadata.getSpanId();
-        linkingMetadata.put("span.id", spanId);
+        linkingMetadata.put(SPAN_ID, spanId);
 
         AgentConfig agentConfig = ServiceFactory.getConfigService().getDefaultAgentConfig();
-        linkingMetadata.put("hostname", getLinkingMetaHostname(agentConfig));
+        linkingMetadata.put(HOSTNAME, getLinkingMetaHostname(agentConfig));
+        try {
+            String entityGuid = ServiceFactory.getRPMService().getEntityGuid();
+            if (!entityGuid.isEmpty()) {
+                linkingMetadata.put(ENTITY_NAME, agentConfig.getApplicationName());
+                linkingMetadata.put(ENTITY_TYPE, "SERVICE");
+                linkingMetadata.put(ENTITY_GUID, entityGuid);
+            }
+        } catch (NullPointerException ignored) {
+            // it's possible to call getLinkingMetadata in the premain before RPMService has been initialized which will NPE
+            Agent.LOG.log(Level.WARNING, "Cannot get entity.guid from getLinkingMetadata() until RPMService has initialized.");
+        }
+
+        return linkingMetadata;
+    }
+
+    /**
+     * Gets a map of agent linking metadata minus entity.type,
+     * entity.name, and any attributes with an empty value.
+     * This subset of linking metadata is added to LogEvents.
+     *
+     * @return Filtered map of agent linking metadata
+     */
+    public static Map<String, String> getLogEventLinkingMetadata() {
+        Map<String, String> linkingMetadata = new ConcurrentHashMap<>();
+
+        TraceMetadata traceMetadata = TraceMetadataImpl.INSTANCE;
+        String traceId = traceMetadata.getTraceId();
+        if (!traceId.isEmpty()) {
+            linkingMetadata.put(TRACE_ID, traceId);
+        }
+
+        String spanId = traceMetadata.getSpanId();
+        if (!spanId.isEmpty()) {
```

Comment:
If there is any chance of null here and you have access to a `StringUtils` in apache commons or guava perhaps, you can do `StringUtils.isBlank()` instead.

Reply:
These won't ever be `null`, the `TraceMetadata` API guarantees that an actual value or an empty `String` will always be returned.

https://github.com/newrelic/newrelic-java-agent/blob/3b934f3ccfbd15437a2f85ef705979148ab5a293/newrelic-api/src/main/java/com/newrelic/api/agent/TraceMetadata.java#L15-L35